File: 28.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0

        def get_next(s):
            nxt = [0]
            i = 1
            now = 0
            while i < len(s):
                if s[i] == s[now]:
                    now += 1
                    nxt.append(now)
                    i += 1
                elif now > 0:
                    now = nxt[now-1]
                else:
                    nxt.append(0)
                    i += 1
            return nxt

        nxt = get_next(needle)
        i = j = 0
        while i < len(haystack):
            if haystack[i] == needle[j]:
                i += 1
                j += 1
                if j == len(needle):
                    return i - j
            elif j > 0:
                j = nxt[j-1]
            else:
                i += 1
        return -1


if __name__ == "__main__":
    s = Solution()
    print(s.strStr('ly', 'l'))

File: 28.go
package leetcode

func strStr(haystack string, needle string) int {
	m, n := len(haystack), len(needle)
	next := make([]int, n)
	i := 1
	now := 0
	for i < n {
		if needle[i] == needle[now] {
			now++
			next[i] = now
			i++
		} else if now > 0 {
			now = next[now-1]
		} else {
			i++
		}
	}
	// fmt.Println(next)
	p := 0
	for q := 0; q < m; {
		if haystack[q] == needle[p] {
			p++
			q++
		} else if p > 0 {
			p = next[p-1]
		} else {
			q++
		}
		// fmt.Println(p, q)
		if p == n {
			return q - n
		}
	}
	return -1
}
